Add an API endpoint to bulk fill rule gaps

denar50 opened the pull request:
## Summary
Original issue: https://github.com/elastic/security-team/issues/10688
This PR adds a new function to the rules client called `bulkFillGapsByRuleIds`, that takes a list of rule ids and a date range, and attempts to fill out the gaps simultaneously, 10 rules at a time (can be customizable).

For all rules we verify if the user has access to write to the gaps object, then for each rule we fetch gaps and trigger the backfilling of its unfilled intervals.

During the execution, we aggregate the errors and the rules that are skipped in lists and return them along with the scheduling outcomes.

### Error handling
There are 2 error types that are handled throughout the scheduling of a gaps backfilling for a rule: 
- A user doesn't have access the rule
- We fail to backfill gaps

### Skipped rules
A rule can be skipped when it doesn't have unfilled gaps in the given time range, or when gaps are found, but they are currently "in progress".

### Endpoint
This PR also updates the existing bulk actions endpoint by adding the action `fill_gaps`. The endpoint for now should limit the amount of rules per call to 100.

The bulk actions endpoint returns a summary of the execution which is a list of counters of rules that succeeded, failed or were skipped.

## How to test?
1. Create several rules that run every 5 seconds and disable them. Leave them disabled for a couple of minutes and then re-enable them in order to create some gaps for it. You can confirm that there are gaps by clicking on the rule, then on the "Execution results" tab, and then scrolling down to "Gaps".
Alternatively you can use [this tool](https://github.com/elastic/security-documents-generator) to create 1 rule (5m) with 1000 gaps:
```
yarn start rules --rules 1 -g 1000 -c -i"5m"
```
2. Get the ids of each rule that you created and call the endpoint to backfill them. You can select a time range that is smaller than the gaps so that you can call it several times. Here is an example of a curl to do this
```
curl -X POST "http://localhost:5601/api/detection_engine/rules/_bulk_action?dry_run=false" \
  -H "Accept: */*" \
  -H "Content-Type: application/json" \
  -H "kbn-version: 9.1.0" \
  -H "kbn-build-number: 9007199254740991" \
  -H "elastic-api-version: 2023-10-31" \
  -H "Authorization: Basic $(echo -n 'elastic:changeme' | base64)" \
  --data-raw '{
    "action": "fill_gaps",
    "ids": ["307bdea8-28be-419f-bb25-dac0024f32af"],
    "fill_gaps": {
      "start_date": "2025-05-09T08:12:09.457Z",
      "end_date": "2025-05-09T09:12:09.457Z"
    }
  }'
```
3. Then you can go to the "Gaps" tab from step 1 and verify that the gaps are being filled. Additionally, you can see that there are manual runs scheduled for the rule in order to fill the gaps.

<!--ONMERGE {"backportTargets":["8.19"]} ONMERGE-->

denar50 committed:
generalize BulkActionSkipResult type

denar50 committed:
complete rules client context mock

denar50 committed:
add bulk fill gaps by rule ids to rules client

denar50 committed:
add a fill_gaps action to the bulk actions endpoint

elasticmachine commented:
<!-- buildkite-pr-bot-trigger-comment -->:robot: Jobs for this PR can be triggered through checkboxes. :construction: 

:information_source: To trigger the CI, please tick the checkbox below :point_down: 

- [ ] <!-- buildkite-pr-bot-checkbox-trigger:kibana-pull-request --> Click to trigger **kibana-pull-request** for this PR!
- [ ] <!-- buildkite-pr-bot-checkbox-trigger:kibana-deploy-project-from-pr --> Click to trigger **kibana-deploy-project-from-pr** for this PR!
- [ ] <!-- buildkite-pr-bot-checkbox-trigger:kibana-deploy-cloud-from-pr --> Click to trigger **kibana-deploy-cloud-from-pr** for this PR!

nkhristinin reviewed (commented):
Hey, thanks for the PR - I make initial high level code review of the main logic.

My main feedback it's about using p-map and combine logic for find and schedule into one operation per rule. 

Let me know if you have any questions!

denar50 committed:
refactor according to draft pr feedback

denar50 commented on x-pack/platform/plugins/shared/alerting/server/application/rule/methods/bulk_fill_gaps_by_rule_ids/get_backfill_schedule_payloads.ts:
```diff
@@ -0,0 +1,117 @@
+  const backfillRequestPayload: ScheduleBackfillParams[0] = {
+    ruleId,
+    ranges: gaps.flatMap((gap) => {
+      const state = gap.getState();
```
Done

denar50 commented on x-pack/platform/plugins/shared/alerting/server/application/rule/methods/bulk_fill_gaps_by_rule_ids/get_backfill_schedule_payloads.ts:
```diff
@@ -0,0 +1,117 @@
+    name: ruleName,
+    gapPagination: { page },
+  } = rule;
+  const { data: gaps, total } = await findGaps(context, {
```
Done

denar50 commented on x-pack/platform/plugins/shared/alerting/server/application/rule/methods/bulk_fill_gaps_by_rule_ids/get_backfill_schedule_payloads.ts:
```diff
@@ -0,0 +1,117 @@
+    start,
+    end,
+    page,
+    statuses: ['partially_filled', 'unfilled'],
```
Done

denar50 commented on x-pack/platform/plugins/shared/alerting/server/application/rule/methods/bulk_fill_gaps_by_rule_ids/get_backfill_schedule_payloads.ts:
```diff
@@ -0,0 +1,117 @@
+  const nextRunRules: GetBackfillSchedulePayloadsResult['next'] = [];
+  const erroredRules: GetBackfillSchedulePayloadsResult['errored'] = [];
+  const payloads: GetBackfillSchedulePayloadsResult['payloads'] = [];
+  await Promise.all(
```
Done

denar50 commented:
# Performance considerations
In order to fill the gaps of a rule, we need to schedule the backfill for all the unfilled intervals of all the gaps for a rule in the given interval.

We ran an experiment where we created 100 rules that run every 5m, each with 1000 gaps that span over 1 month of unfilled gaps. Then we attempt to backfill all those gaps by processing 10 rules concurrently at a given time. 

Under these circumstances, the existing logic takes around 108s to schedule the backfilling of all the gaps.

We noticed that updating the state of the gaps is the biggest bottleneck here, since disabling that logic reduces the execution time to approximately 20s which represents a reduction of 80% of the time.

denar50 committed:
avoid refetching gaps when backfilling and limit amount of backfilled gaps to 1000

denar50 committed:
Add parameters validation to bulkFillGapsByRuleIds

denar50 committed:
batch run authorization checks on rules

denar50 committed:
Add integration tests

denar50 committed:
fix linting error

kibanamachine committed:
[CI] Auto-commit changed files from 'yarn openapi:bundle'

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/305933)
* Commit: fc33e8707708794ecdda3092a3a237b536f7ac74

### Failed CI Steps
* [Check Types](https://buildkite.com/elastic/kibana-pull-request/builds/305933#0197414c-de7a-465c-9960-c0d4e6f76741)

### History
* :broken_heart: [Build #305927](https://buildkite.com/elastic/kibana-pull-request/builds/305927) failed 372a75e088ccaf9da575cb29266d20ccbe9e5786
* :broken_heart: [Build #305821](https://buildkite.com/elastic/kibana-pull-request/builds/305821) failed 63df54d9fb59a704f8452c65eb19cec04b7e95fc
* :broken_heart: [Build #305805](https://buildkite.com/elastic/kibana-pull-request/builds/305805) failed e6b4200c25f53e1f8ad907a0ff5e6a3ceb05f829

cc @denar50

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/305933","number":305933,"commit":"fc33e8707708794ecdda3092a3a237b536f7ac74"},{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/305927","number":305927,"commit":"372a75e088ccaf9da575cb29266d20ccbe9e5786"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/305821","number":305821,"commit":"63df54d9fb59a704f8452c65eb19cec04b7e95fc"},{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/305805","number":305805,"commit":"e6b4200c25f53e1f8ad907a0ff5e6a3ceb05f829"}],"number":305933}
buildkite-pr-comment-->

kibanamachine committed:
[CI] Auto-commit changed files from 'make api-docs'

elasticmachine commented:
Pinging @elastic/security-detection-engine (Team:Detection Engine)

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/306605)
* Commit: 862f6251640d8383f42e5a117ba37e646a09a0c8

### Failed CI Steps
* [Check Types](https://buildkite.com/elastic/kibana-pull-request/builds/306605#019754c1-92e8-4b4f-9191-751ed2c73b0a)

### History
* :broken_heart: [Build #305937](https://buildkite.com/elastic/kibana-pull-request/builds/305937) failed 31d5be9861be1cce16c86116e32c08ab23776fa2
* :broken_heart: [Build #305933](https://buildkite.com/elastic/kibana-pull-request/builds/305933) failed fc33e8707708794ecdda3092a3a237b536f7ac74
* :broken_heart: [Build #305927](https://buildkite.com/elastic/kibana-pull-request/builds/305927) failed 372a75e088ccaf9da575cb29266d20ccbe9e5786
* :broken_heart: [Build #305821](https://buildkite.com/elastic/kibana-pull-request/builds/305821) failed 63df54d9fb59a704f8452c65eb19cec04b7e95fc
* :broken_heart: [Build #305805](https://buildkite.com/elastic/kibana-pull-request/builds/305805) failed e6b4200c25f53e1f8ad907a0ff5e6a3ceb05f829

cc @denar50

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/306605","number":306605,"commit":"862f6251640d8383f42e5a117ba37e646a09a0c8"},{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/305937","number":305937,"commit":"31d5be9861be1cce16c86116e32c08ab23776fa2"},{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/305933","number":305933,"commit":"fc33e8707708794ecdda3092a3a237b536f7ac74"},{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/305927","number":305927,"commit":"372a75e088ccaf9da575cb29266d20ccbe9e5786"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/305821","number":305821,"commit":"63df54d9fb59a704f8452c65eb19cec04b7e95fc"},{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/305805","number":305805,"commit":"e6b4200c25f53e1f8ad907a0ff5e6a3ceb05f829"}],"number":306605}
buildkite-pr-comment-->

denar50 committed:
refactor bulk action skip result type

elasticmachine commented:
## :broken_heart: Build Failed
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/306636)
* Commit: e3794bc390f522b3edd2466fd6cf524aa8937b2c

### Failed CI Steps
* [Check Types](https://buildkite.com/elastic/kibana-pull-request/builds/306636#019754f8-841d-4b84-8f8f-4d3990aa98eb)

### History
* :broken_heart: [Build #306605](https://buildkite.com/elastic/kibana-pull-request/builds/306605) failed 862f6251640d8383f42e5a117ba37e646a09a0c8
* :broken_heart: [Build #305937](https://buildkite.com/elastic/kibana-pull-request/builds/305937) failed 31d5be9861be1cce16c86116e32c08ab23776fa2
* :broken_heart: [Build #305933](https://buildkite.com/elastic/kibana-pull-request/builds/305933) failed fc33e8707708794ecdda3092a3a237b536f7ac74
* :broken_heart: [Build #305927](https://buildkite.com/elastic/kibana-pull-request/builds/305927) failed 372a75e088ccaf9da575cb29266d20ccbe9e5786
* :broken_heart: [Build #305821](https://buildkite.com/elastic/kibana-pull-request/builds/305821) failed 63df54d9fb59a704f8452c65eb19cec04b7e95fc

cc @denar50

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/306636","number":306636,"commit":"e3794bc390f522b3edd2466fd6cf524aa8937b2c"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/306605","number":306605,"commit":"862f6251640d8383f42e5a117ba37e646a09a0c8"},{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/305937","number":305937,"commit":"31d5be9861be1cce16c86116e32c08ab23776fa2"},{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/305933","number":305933,"commit":"fc33e8707708794ecdda3092a3a237b536f7ac74"},{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/305927","number":305927,"commit":"372a75e088ccaf9da575cb29266d20ccbe9e5786"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/305821","number":305821,"commit":"63df54d9fb59a704f8452c65eb19cec04b7e95fc"}],"number":306636}
buildkite-pr-comment-->

denar50 committed:
add frontend translations to fix the type errors of the BulkAction

This copy was going to be added in a follow up PR with the frontend changes, but this particular change had to be added because the server type is being using here.

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/306716)
* Commit: 3c80bc6bb2159012963536078d931100546c02aa

### Failed CI Steps
* [Serverless Detection Engine - Security Solution Cypress Tests #5](https://buildkite.com/elastic/kibana-pull-request/builds/306716#019755e9-4c7a-4461-8967-95d219e9d5b1)

### Test Failures
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/306716#019755e9-4c7a-4461-8967-95d219e9d5b1) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/306716/jobs/019755e9-4c7a-4461-8967-95d219e9d5b1/artifacts/0197561f-15d2-4af0-a32f-96b9bcf03f4e) Serverless Detection Engine - Security Solution Cypress Tests #<span></span>5 / Manual rule run schedule enable rules and show partial error for disabled rules when all rules are selected schedule enable rules and show partial error for disabled rules when all rules are selected

### History
* :broken_heart: [Build #306636](https://buildkite.com/elastic/kibana-pull-request/builds/306636) failed e3794bc390f522b3edd2466fd6cf524aa8937b2c
* :broken_heart: [Build #306605](https://buildkite.com/elastic/kibana-pull-request/builds/306605) failed 862f6251640d8383f42e5a117ba37e646a09a0c8
* :broken_heart: [Build #305937](https://buildkite.com/elastic/kibana-pull-request/builds/305937) failed 31d5be9861be1cce16c86116e32c08ab23776fa2
* :broken_heart: [Build #305933](https://buildkite.com/elastic/kibana-pull-request/builds/305933) failed fc33e8707708794ecdda3092a3a237b536f7ac74
* :broken_heart: [Build #305927](https://buildkite.com/elastic/kibana-pull-request/builds/305927) failed 372a75e088ccaf9da575cb29266d20ccbe9e5786
* :broken_heart: [Build #305821](https://buildkite.com/elastic/kibana-pull-request/builds/305821) failed 63df54d9fb59a704f8452c65eb19cec04b7e95fc

cc @denar50

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/306716","number":306716,"commit":"3c80bc6bb2159012963536078d931100546c02aa"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/306636","number":306636,"commit":"e3794bc390f522b3edd2466fd6cf524aa8937b2c"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/306605","number":306605,"commit":"862f6251640d8383f42e5a117ba37e646a09a0c8"},{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/305937","number":305937,"commit":"31d5be9861be1cce16c86116e32c08ab23776fa2"},{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/305933","number":305933,"commit":"fc33e8707708794ecdda3092a3a237b536f7ac74"},{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/305927","number":305927,"commit":"372a75e088ccaf9da575cb29266d20ccbe9e5786"}],"number":306716}
buildkite-pr-comment-->

denar50 committed:
fix error propagation in bulkScheduleBackfill
the logic to process the backfill results was abstracted and it no longer took into account the validation errors that occurred before schedulign the backfilling.

denar50 committed:
fix snapshot tests
since there is a new bulk action type `fill_gaps`, its validation errors include the amount of errors in the returned array.

marshallmain reviewed (commented):
Notes:
- `calculateGapStateFromAllBackfills` (https://github.com/elastic/kibana/blob/v9.0.2/x-pack/platform/plugins/shared/alerting/server/lib/rule_gaps/update/calculate_gaps_state.ts#L43-L49) reassigns `gap` and returns it, but the return value is not used (https://github.com/elastic/kibana/blob/v9.0.2/x-pack/platform/plugins/shared/alerting/server/lib/rule_gaps/update/update_gaps.ts#L73)?

Can you also update the PR title to be more descriptive and remove "Security team 10688"?

elasticmachine commented:
## :green_heart: Build Succeeded
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/307019)
* Commit: 175c95569c49b2e06238069488f7ee07a9eb7c3c

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
#### Async chunks
> Total size of all lazy-loaded chunks that will be downloaded as the user navigates the app


| id | [before](https://github.com/elastic/kibana/commit/043ebe521e43b95d70dba1785096af7962ba63e1) | [after](https://github.com/elastic/kibana/commit/175c95569c49b2e06238069488f7ee07a9eb7c3c) | diff |
| --- | --- | --- | --- |
| `securitySolution` | 9.3MB | 9.3MB | +1.3KB |

#### Public APIs missing exports
> Total count of every type that is part of your API that should be exported but is not. This will cause broken links in the API documentation system. Target amount is 0. `Run node scripts/build_api_docs --plugin [yourplugin] --stats exports` for more detailed information.


| id | [before](https://github.com/elastic/kibana/commit/043ebe521e43b95d70dba1785096af7962ba63e1) | [after](https://github.com/elastic/kibana/commit/175c95569c49b2e06238069488f7ee07a9eb7c3c) | diff |
| --- | --- | --- | --- |
| `alerting` | 53 | 55 | +2 |



### History
* :broken_heart: [Build #306716](https://buildkite.com/elastic/kibana-pull-request/builds/306716) failed 3c80bc6bb2159012963536078d931100546c02aa
* :broken_heart: [Build #306636](https://buildkite.com/elastic/kibana-pull-request/builds/306636) failed e3794bc390f522b3edd2466fd6cf524aa8937b2c
* :broken_heart: [Build #306605](https://buildkite.com/elastic/kibana-pull-request/builds/306605) failed 862f6251640d8383f42e5a117ba37e646a09a0c8
* :broken_heart: [Build #305937](https://buildkite.com/elastic/kibana-pull-request/builds/305937) failed 31d5be9861be1cce16c86116e32c08ab23776fa2
* :broken_heart: [Build #305933](https://buildkite.com/elastic/kibana-pull-request/builds/305933) failed fc33e8707708794ecdda3092a3a237b536f7ac74

cc @denar50

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/307019","number":307019,"commit":"175c95569c49b2e06238069488f7ee07a9eb7c3c"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/306716","number":306716,"commit":"3c80bc6bb2159012963536078d931100546c02aa"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/306636","number":306636,"commit":"e3794bc390f522b3edd2466fd6cf524aa8937b2c"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/306605","number":306605,"commit":"862f6251640d8383f42e5a117ba37e646a09a0c8"},{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/305937","number":305937,"commit":"31d5be9861be1cce16c86116e32c08ab23776fa2"},{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/305933","number":305933,"commit":"fc33e8707708794ecdda3092a3a237b536f7ac74"}],"number":307019}
buildkite-pr-comment-->

denar50 commented on x-pack/platform/plugins/shared/alerting/server/lib/rule_gaps/process_all_gaps_in_time_range.ts:
```diff
@@ -0,0 +1,113 @@
+// Circuit breaker to prevent infinite loops
+// It should be enough to update 50,000,000 gaps
+// 100000 * 500 = 50,000,000 millions gaps
+const DEFAULT_MAX_ITERATIONS = 100000;
```
I believe the intention was to pick a very high number, but maybe @nkhristinin can add more information as to why?

denar50 committed:
Fetch and process gaps in batches

denar50 committed:
add integration tests for dry run

denar50 committed:
Rename variable

denar50 commented on x-pack/platform/plugins/shared/alerting/server/application/rule/methods/bulk_fill_gaps_by_rule_ids/bulk_fill_gaps_by_rule_ids.ts:
```diff
@@ -0,0 +1,137 @@
+    options?.maxBackfillConcurrency ?? DEFAULT_MAX_BACKFILL_CONCURRENCY;
+
+  // Make sure user has access to these rules
+  const alertTypes = mapValues(
```
Done

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/307393)
* Commit: 1f762d599c5b61c8236d8fc811dd000ecff7333d

### Failed CI Steps
* [Defend Workflows Cypress Tests #7](https://buildkite.com/elastic/kibana-pull-request/builds/307393#01975ee5-90eb-4fe3-99fa-66b6929e788c)

### History
* :broken_heart: [Build #307376](https://buildkite.com/elastic/kibana-pull-request/builds/307376) failed 42e5d95f2d0a0dcd9a3b690d8f0dd630fbbe4420
* :green_heart: [Build #307019](https://buildkite.com/elastic/kibana-pull-request/builds/307019) succeeded 175c95569c49b2e06238069488f7ee07a9eb7c3c
* :broken_heart: [Build #306716](https://buildkite.com/elastic/kibana-pull-request/builds/306716) failed 3c80bc6bb2159012963536078d931100546c02aa
* :broken_heart: [Build #306636](https://buildkite.com/elastic/kibana-pull-request/builds/306636) failed e3794bc390f522b3edd2466fd6cf524aa8937b2c
* :broken_heart: [Build #306605](https://buildkite.com/elastic/kibana-pull-request/builds/306605) failed 862f6251640d8383f42e5a117ba37e646a09a0c8

cc @denar50

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/307393","number":307393,"commit":"1f762d599c5b61c8236d8fc811dd000ecff7333d"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/307376","number":307376,"commit":"42e5d95f2d0a0dcd9a3b690d8f0dd630fbbe4420"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/307019","number":307019,"commit":"175c95569c49b2e06238069488f7ee07a9eb7c3c"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/306716","number":306716,"commit":"3c80bc6bb2159012963536078d931100546c02aa"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/306636","number":306636,"commit":"e3794bc390f522b3edd2466fd6cf524aa8937b2c"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/306605","number":306605,"commit":"862f6251640d8383f42e5a117ba37e646a09a0c8"}],"number":307393}
buildkite-pr-comment-->

kibanamachine committed:
[CI] Auto-commit changed files from 'yarn openapi:bundle'

kibanamachine committed:
[CI] Auto-commit changed files from 'make api-docs'

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/307631)
* Commit: b55ae7a2419104f442b64dcf793d1bba25aeacd3

### Failed CI Steps
* [Entity Analytics - Security Solution Cypress Tests #2](https://buildkite.com/elastic/kibana-pull-request/builds/307631#019760d4-94de-4493-a0bd-03443854afea)

### History
* :yellow_heart: [Build #307393](https://buildkite.com/elastic/kibana-pull-request/builds/307393) was flaky 1f762d599c5b61c8236d8fc811dd000ecff7333d
* :broken_heart: [Build #307376](https://buildkite.com/elastic/kibana-pull-request/builds/307376) failed 42e5d95f2d0a0dcd9a3b690d8f0dd630fbbe4420
* :green_heart: [Build #307019](https://buildkite.com/elastic/kibana-pull-request/builds/307019) succeeded 175c95569c49b2e06238069488f7ee07a9eb7c3c
* :broken_heart: [Build #306716](https://buildkite.com/elastic/kibana-pull-request/builds/306716) failed 3c80bc6bb2159012963536078d931100546c02aa
* :broken_heart: [Build #306636](https://buildkite.com/elastic/kibana-pull-request/builds/306636) failed e3794bc390f522b3edd2466fd6cf524aa8937b2c

cc @denar50

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/307631","number":307631,"commit":"b55ae7a2419104f442b64dcf793d1bba25aeacd3"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/307393","number":307393,"commit":"1f762d599c5b61c8236d8fc811dd000ecff7333d"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/307376","number":307376,"commit":"42e5d95f2d0a0dcd9a3b690d8f0dd630fbbe4420"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/307019","number":307019,"commit":"175c95569c49b2e06238069488f7ee07a9eb7c3c"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/306716","number":306716,"commit":"3c80bc6bb2159012963536078d931100546c02aa"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/306636","number":306636,"commit":"e3794bc390f522b3edd2466fd6cf524aa8937b2c"}],"number":307631}
buildkite-pr-comment-->

nkhristinin reviewed (approved):
Great work!  Tested locally, and it's working.

Left some minor comments! but LGTM!

denar50 commented on x-pack/platform/plugins/shared/alerting/server/application/rule/methods/bulk_fill_gaps_by_rule_ids/schemas/params.ts:
```diff
@@ -0,0 +1,66 @@
+      return end - start >= FIVE_MINUTES_MS;
+    },
+    {
+      message: 'end must be at least 5 minutes after start date.',
```
Hmm, thought that would be a reasonable interval for a rule execution, but happy to change it to a smaller value or just to verify that `end` has to be greater than `start`.

denar50 committed:
Export and use PROCESS_GAPS_DEFAULT_PAGE_SIZE in update gaps logic

denar50 committed:
Merge branch 'main' into security-team-10688-bulk-fill-gaps-for-rule-be

denar50 committed:
Fix span noise

denar50 committed:
Merge branch 'main' into security-team-10688-bulk-fill-gaps-for-rule-be

jkelas reviewed (commented):
The code looks very good. 
I tested that it works as described. I created 5 rules with gaps and made a request to the bulk_action endpoint to fill the gaps. All rules succeded. I am pasting only the summary below.
I synced with the author and we discussed the changes, we walked through the code and through Kibana UI. Everything looked OK. I could see the gaps as filled and also I could observe scheduled runs to gradually fill the gaps. 

```
        "summary": {
            "failed": 0,
            "succeeded": 5,
            "skipped": 0,
            "total": 5
        }
```

I only left some comments for some improvements. @denar50, when you address the comments I will approve the PR.

elasticmachine commented:
## :green_heart: Build Succeeded
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/308234)
* Commit: d2366c324f016b1a433350f3b8d1f69805b75336

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
#### Async chunks
> Total size of all lazy-loaded chunks that will be downloaded as the user navigates the app


| id | [before](https://github.com/elastic/kibana/commit/d5d37c7b2ab9b84d4aafe80b6fdc572b8f95035b) | [after](https://github.com/elastic/kibana/commit/d2366c324f016b1a433350f3b8d1f69805b75336) | diff |
| --- | --- | --- | --- |
| `securitySolution` | 9.3MB | 9.3MB | +1.3KB |

#### Public APIs missing exports
> Total count of every type that is part of your API that should be exported but is not. This will cause broken links in the API documentation system. Target amount is 0. `Run node scripts/build_api_docs --plugin [yourplugin] --stats exports` for more detailed information.


| id | [before](https://github.com/elastic/kibana/commit/d5d37c7b2ab9b84d4aafe80b6fdc572b8f95035b) | [after](https://github.com/elastic/kibana/commit/d2366c324f016b1a433350f3b8d1f69805b75336) | diff |
| --- | --- | --- | --- |
| `alerting` | 54 | 56 | +2 |



### History
* :green_heart: [Build #307876](https://buildkite.com/elastic/kibana-pull-request/builds/307876) succeeded 6bc9576cb762ca802f683263470fff2a0f986c2a
* :yellow_heart: [Build #307631](https://buildkite.com/elastic/kibana-pull-request/builds/307631) was flaky b55ae7a2419104f442b64dcf793d1bba25aeacd3
* :yellow_heart: [Build #307393](https://buildkite.com/elastic/kibana-pull-request/builds/307393) was flaky 1f762d599c5b61c8236d8fc811dd000ecff7333d
* :broken_heart: [Build #307376](https://buildkite.com/elastic/kibana-pull-request/builds/307376) failed 42e5d95f2d0a0dcd9a3b690d8f0dd630fbbe4420
* :green_heart: [Build #307019](https://buildkite.com/elastic/kibana-pull-request/builds/307019) succeeded 175c95569c49b2e06238069488f7ee07a9eb7c3c
* :broken_heart: [Build #306716](https://buildkite.com/elastic/kibana-pull-request/builds/306716) failed 3c80bc6bb2159012963536078d931100546c02aa

cc @denar50

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/308234","number":308234,"commit":"d2366c324f016b1a433350f3b8d1f69805b75336"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/307876","number":307876,"commit":"6bc9576cb762ca802f683263470fff2a0f986c2a"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/307631","number":307631,"commit":"b55ae7a2419104f442b64dcf793d1bba25aeacd3"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/307393","number":307393,"commit":"1f762d599c5b61c8236d8fc811dd000ecff7333d"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/307376","number":307376,"commit":"42e5d95f2d0a0dcd9a3b690d8f0dd630fbbe4420"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/307019","number":307019,"commit":"175c95569c49b2e06238069488f7ee07a9eb7c3c"}],"number":308234}
buildkite-pr-comment-->

denar50 committed:
losen the date range validation to only require the end date to be greater than the start date

denar50 committed:
improve open api copy

denar50 commented on x-pack/solutions/security/plugins/security_solution/common/api/detection_engine/rule_management/bulk_actions/bulk_actions_route.schema.yaml:
```diff
@@ -1245,6 +1265,31 @@ components:
+              enum: [fill_gaps]
+            fill_gaps:
+              type: object
+              description: Object that describes applying a manual gap fill action for the specified range.
```
Done

denar50 committed:
add a validation function specific for gap filling

denar50 committed:
refactor validation logic in bulkScheduleRuleGapFilling

denar50 committed:
small nit refactoring in test

denar50 commented on x-pack/test/security_solution_api_integration/test_suites/detections_response/rules_management/rule_bulk_actions/trial_license_complete_tier/perform_bulk_action.ts:
```diff
@@ -2612,6 +2615,299 @@ export default ({ getService }: FtrProviderContext): void => {
+          total: 2,
+        });
+
+        const expectedUpdatedRules = Object.values(generatedGapEvents)
```
done

kibanamachine committed:
[CI] Auto-commit changed files from 'yarn openapi:bundle'

jkelas reviewed (approved):
My comments have been addressed, so I am approving. Well done.

kibanamachine committed:
[CI] Auto-commit changed files from 'make api-docs'

denar50 committed:
increase the limit of max gaps per rule when the endpoint is called with one rule id

denar50 committed:
fix error message in integration test

denar50 committed:
fix integration test

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/308505)
* Commit: cbaf2323fe6c87140f8a3fd74c55e97a0fc15c55

### Failed CI Steps
* [Osquery Cypress Tests on Serverless #2](https://buildkite.com/elastic/kibana-pull-request/builds/308505#01976b49-4200-4f1e-84c8-a359c2518bbb)

### History
* :broken_heart: [Build #308475](https://buildkite.com/elastic/kibana-pull-request/builds/308475) failed 9247253ed3b2e86b1dcb09daba15752adead9380
* :broken_heart: [Build #308364](https://buildkite.com/elastic/kibana-pull-request/builds/308364) failed 437009ce6c0e8836da8db650dbe65370d676a7b1
* :broken_heart: [Build #308302](https://buildkite.com/elastic/kibana-pull-request/builds/308302) failed f78d8a8a2132d63535aa60edddef86f005ce51dd
* :green_heart: [Build #308234](https://buildkite.com/elastic/kibana-pull-request/builds/308234) succeeded d2366c324f016b1a433350f3b8d1f69805b75336
* :green_heart: [Build #307876](https://buildkite.com/elastic/kibana-pull-request/builds/307876) succeeded 6bc9576cb762ca802f683263470fff2a0f986c2a

cc @denar50

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/308505","number":308505,"commit":"cbaf2323fe6c87140f8a3fd74c55e97a0fc15c55"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/308475","number":308475,"commit":"9247253ed3b2e86b1dcb09daba15752adead9380"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/308364","number":308364,"commit":"437009ce6c0e8836da8db650dbe65370d676a7b1"},{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/308302","number":308302,"commit":"f78d8a8a2132d63535aa60edddef86f005ce51dd"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/308234","number":308234,"commit":"d2366c324f016b1a433350f3b8d1f69805b75336"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/307876","number":307876,"commit":"6bc9576cb762ca802f683263470fff2a0f986c2a"}],"number":308505}
buildkite-pr-comment-->

denar50 committed:
Merge branch 'main' into security-team-10688-bulk-fill-gaps-for-rule-be

elasticmachine commented:
## :green_heart: Build Succeeded
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/308602)
* Commit: 51037dfe66d684345dc1969e7a0e209c3229aa98

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
#### Async chunks
> Total size of all lazy-loaded chunks that will be downloaded as the user navigates the app


| id | [before](https://github.com/elastic/kibana/commit/951bc75b27116e5be51f1b0a8d262e7e89d58009) | [after](https://github.com/elastic/kibana/commit/51037dfe66d684345dc1969e7a0e209c3229aa98) | diff |
| --- | --- | --- | --- |
| `securitySolution` | 9.3MB | 9.3MB | +1.3KB |

#### Public APIs missing exports
> Total count of every type that is part of your API that should be exported but is not. This will cause broken links in the API documentation system. Target amount is 0. `Run node scripts/build_api_docs --plugin [yourplugin] --stats exports` for more detailed information.


| id | [before](https://github.com/elastic/kibana/commit/951bc75b27116e5be51f1b0a8d262e7e89d58009) | [after](https://github.com/elastic/kibana/commit/51037dfe66d684345dc1969e7a0e209c3229aa98) | diff |
| --- | --- | --- | --- |
| `alerting` | 54 | 57 | +3 |



### History
* :broken_heart: [Build #308505](https://buildkite.com/elastic/kibana-pull-request/builds/308505) failed cbaf2323fe6c87140f8a3fd74c55e97a0fc15c55
* :broken_heart: [Build #308475](https://buildkite.com/elastic/kibana-pull-request/builds/308475) failed 9247253ed3b2e86b1dcb09daba15752adead9380
* :broken_heart: [Build #308364](https://buildkite.com/elastic/kibana-pull-request/builds/308364) failed 437009ce6c0e8836da8db650dbe65370d676a7b1
* :broken_heart: [Build #308302](https://buildkite.com/elastic/kibana-pull-request/builds/308302) failed f78d8a8a2132d63535aa60edddef86f005ce51dd
* :green_heart: [Build #308234](https://buildkite.com/elastic/kibana-pull-request/builds/308234) succeeded d2366c324f016b1a433350f3b8d1f69805b75336

cc @denar50

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/308602","number":308602,"commit":"51037dfe66d684345dc1969e7a0e209c3229aa98"},{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/308505","number":308505,"commit":"cbaf2323fe6c87140f8a3fd74c55e97a0fc15c55"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/308475","number":308475,"commit":"9247253ed3b2e86b1dcb09daba15752adead9380"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/308364","number":308364,"commit":"437009ce6c0e8836da8db650dbe65370d676a7b1"},{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/308302","number":308302,"commit":"f78d8a8a2132d63535aa60edddef86f005ce51dd"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/308234","number":308234,"commit":"d2366c324f016b1a433350f3b8d1f69805b75336"}],"number":308602}
buildkite-pr-comment-->

ymao1 commented on x-pack/platform/plugins/shared/alerting/server/application/rule/methods/bulk_fill_gaps_by_rule_ids/bulk_fill_gaps_by_rule_ids.ts:
```diff
@@ -0,0 +1,97 @@
+    options?.maxBackfillConcurrency ?? DEFAULT_MAX_BACKFILL_CONCURRENCY;
+
+  // Make sure user has access to these rules
+  const rulesByAlertType = mapValues(
```
```suggestion
  const rulesByRuleType = mapValues(
```

denar50 replied:
Done

ymao1 commented on x-pack/platform/plugins/shared/alerting/server/application/rule/methods/bulk_fill_gaps_by_rule_ids/bulk_fill_gaps_by_rule_ids.ts:
```diff
@@ -0,0 +1,97 @@
+  const rulesByAlertType = mapValues(
+    groupBy(rules, (rule) => `${rule.alertTypeId}<>${rule.consumer}`),
+    (groupedRules) => ({
+      alertTypeId: groupedRules[0].alertTypeId,
```
```suggestion
      ruleTypeId: groupedRules[0].alertTypeId,
```

I know the field is called `alertTypeId` but it's a legacy naming that is difficult to change since it's a saved object field. In our code, we try to use the correct terminology 🙈

denar50 replied:
Done

ymao1 commented on x-pack/platform/plugins/shared/alerting/server/application/rule/methods/bulk_fill_gaps_by_rule_ids/batch_backfill_rule_gaps.ts:
```diff
@@ -0,0 +1,107 @@
+}
+
+type BatchBackfillScheduleRuleGapsResult =
+  | { outcome: 'backfilled' }
```
nit: could we make these outcomes an enum instead of hardcoding the strings?

ymao1 commented on x-pack/platform/plugins/shared/alerting/server/application/rule/methods/bulk_fill_gaps_by_rule_ids/batch_backfill_rule_gaps.ts:
```diff
@@ -0,0 +1,107 @@
+  context: RulesClientContext,
+  { rule, range, maxGapCountPerRule }: BatchBackfillRuleGapsParams
+): Promise<BatchBackfillScheduleRuleGapsResult> => {
+  const logger = context.logger;
```
nit: would it be useful to add context to this logger so it's `log.logger` is  `plugins.alerting.gaps`. or add some tags so it's easier to search for?

ymao1 commented on x-pack/platform/plugins/shared/alerting/server/lib/rule_gaps/process_all_gaps_in_time_range.ts:
```diff
@@ -0,0 +1,113 @@
+        gapsToProcess = gapsToProcess.slice(0, gaps.length - offset);
+      }
+
+      await processGapsBatch(gapsToProcess);
```
should we check for `gapsToProcess.length > 0` and skip calling `processGapsBatch` if empty array?

denar50 replied:
Done

ymao1 reviewed (approved):
Response ops changes LGTM. Code review only of alerting code. Left some nits

marshallmain commented on x-pack/platform/plugins/shared/alerting/server/application/rule/methods/bulk_fill_gaps_by_rule_ids/batch_backfill_rule_gaps.test.ts:
```diff
@@ -0,0 +1,199 @@
+
+const scheduleBackfillMock = scheduleBackfill as jest.Mock;
+
+describe('batchBackfillRuleGaps', () => {
```
These tests are overly complex, in part due to `processGapsBatch` being defined inside `batchBackfillRuleGaps` and thus not being easily testable in isolation. IMO we should move `processGapsBatch` outside of `batchBackfillRuleGaps` and focus each test on the logic of one function or the other.

denar50 replied:
I have moved `processGapsBatch` to its own file and added tests for it.

marshallmain commented on x-pack/platform/plugins/shared/alerting/server/lib/rule_gaps/process_all_gaps_in_time_range.ts:
```diff
@@ -0,0 +1,113 @@
+  start: string;
+  end: string;
+  statuses?: GapStatus[];
+  options?: {
```
I'm not sure I understand - for testing, can't we programmatically make a large array of gaps to trigger any `maxFetchedGaps` logic? And we're also not using `pageSize` in non-test code afaict. So it seems like we should just have a static page size set, pick either maxIterations or maxFetchedGaps as the circuit breaker, and allow that circuit breaker to be passed in. Then the unit tests should make more data if needed - 1000 items isn't enough to be a scale concern in a unit test imo.

denar50 replied:
Gotcha. I have removed the options added for testing and instead programmatically generated gaps to test the case when we reach the max number of iterations.

marshallmain reviewed (commented):
Looks good overall, after these comments are addressed I think it'll be good to merge

denar50 committed:
Merge branch 'main' into security-team-10688-bulk-fill-gaps-for-rule-be

denar50 committed:
PR improvements:
- move process gaps batch to its own file and add tests specific to it
- use enums for the sum types used to represent the outcomes of the bulk fil gaps logic

denar50 committed:
fix typo

denar50 commented on x-pack/test/security_solution_api_integration/test_suites/detections_response/rules_management/rule_bulk_actions/trial_license_complete_tier/perform_bulk_action.ts:
```diff
@@ -2612,6 +2615,299 @@ export default ({ getService }: FtrProviderContext): void => {
+      });
+
+      it('should return 500 error if some rules do not exist', async () => {
+        const existantRules = Object.keys(generatedGapEvents);
```
fixed

denar50 commented on x-pack/platform/plugins/shared/alerting/server/application/rule/methods/bulk_fill_gaps_by_rule_ids/batch_backfill_rule_gaps.test.ts:
```diff
@@ -0,0 +1,199 @@
+      });
+    });
+
+    it('should trigger the backfilling of each fetched gap batch', () => {
```
Done

denar50 commented on x-pack/platform/plugins/shared/alerting/server/application/rule/methods/bulk_fill_gaps_by_rule_ids/batch_backfill_rule_gaps.ts:
```diff
@@ -0,0 +1,107 @@
+      const backfillError = results[0].error;
+      throw new Error(backfillError.message);
+    } else {
+      hasBeenBackfilled = true;
```
Now `processGapsBatch` returns a boolean indicating whether the batch was scheduled or not. `batchBackfillRuleGaps` now returns an array with the results of each call to `processGapsBatch`. Therefore `hasBeenBackfilled` is now `true` when there is at least one `processGapsBatch` call returns true.

denar50 committed:
remove unnecessary options for processAllGapsInTimeRange and improve its tests

denar50 committed:
test the case when findGapsSearchAfter returns an empty array

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/309954)
* Commit: d9177f09940603f0b41762f3ece68ed8d86b0f5f

### Failed CI Steps
* [Check Types](https://buildkite.com/elastic/kibana-pull-request/builds/309954#01978330-ef25-4c1d-a624-f8eb7124133d)

### History
* :broken_heart: [Build #309940](https://buildkite.com/elastic/kibana-pull-request/builds/309940) failed 48d16fff4f81298f0c8680d923faf600f0f84771
* :broken_heart: [Build #309881](https://buildkite.com/elastic/kibana-pull-request/builds/309881) failed 83e6af48c11b84960577445324bafeac245319c8
* :broken_heart: [Build #309731](https://buildkite.com/elastic/kibana-pull-request/builds/309731) failed f571801f381aa00fcfb651c6ab1f1f49acb09153
* :green_heart: [Build #308602](https://buildkite.com/elastic/kibana-pull-request/builds/308602) succeeded 51037dfe66d684345dc1969e7a0e209c3229aa98
* :broken_heart: [Build #308505](https://buildkite.com/elastic/kibana-pull-request/builds/308505) failed cbaf2323fe6c87140f8a3fd74c55e97a0fc15c55
* :broken_heart: [Build #308475](https://buildkite.com/elastic/kibana-pull-request/builds/308475) failed 9247253ed3b2e86b1dcb09daba15752adead9380

cc @denar50

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/309954","number":309954,"commit":"d9177f09940603f0b41762f3ece68ed8d86b0f5f"},{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/309940","number":309940,"commit":"48d16fff4f81298f0c8680d923faf600f0f84771"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/309881","number":309881,"commit":"83e6af48c11b84960577445324bafeac245319c8"},{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/309731","number":309731,"commit":"f571801f381aa00fcfb651c6ab1f1f49acb09153"},{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/308602","number":308602,"commit":"51037dfe66d684345dc1969e7a0e209c3229aa98"},{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/308505","number":308505,"commit":"cbaf2323fe6c87140f8a3fd74c55e97a0fc15c55"}],"number":309954}
buildkite-pr-comment-->

denar50 committed:
reuse validateBackfillSchedule

denar50 commented on x-pack/platform/plugins/shared/alerting/server/application/rule/methods/bulk_fill_gaps_by_rule_ids/schemas/params.ts:
```diff
@@ -0,0 +1,65 @@
+  message: 'Invalid ISO date string.',
+});
+
+const dateRangeSchema = z
```
Done

denar50 committed:
rename variable and explicitly check for BulkFillGapsScheduleResult.SKIPPED

denar50 committed:
add max value to maxBackfillConcurrency

denar50 commented on x-pack/platform/plugins/shared/alerting/server/application/rule/methods/bulk_fill_gaps_by_rule_ids/bulk_fill_gaps_by_rule_ids.ts:
```diff
@@ -0,0 +1,97 @@
+  const backfilled: BulkFillGapsByRuleIdsParams['rules'] = [];
+  const eventLogClient = await context.getEventLogClient();
+  const maxBackfillConcurrency =
+    options?.maxBackfillConcurrency ?? DEFAULT_MAX_BACKFILL_CONCURRENCY;
```
Done

denar50 committed:
add a test to make sure that ensureAuthorized gets called once per combination of ruleTypeId and consumer

denar50 committed:
fix build errors

denar50 committed:
Merge branch 'main' into security-team-10688-bulk-fill-gaps-for-rule-be

denar50 committed:
add context to the logger in batchBackfillRuleGaps

denar50 committed:
Merge branch 'main' into security-team-10688-bulk-fill-gaps-for-rule-be

elasticmachine commented:
## :green_heart: Build Succeeded
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/310058)
* Commit: 9806b41fa9d260b42f7ec81e8927caeadf949066

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
#### Async chunks
> Total size of all lazy-loaded chunks that will be downloaded as the user navigates the app


| id | [before](https://github.com/elastic/kibana/commit/32e337401350aa1c34a30ceacc92b642d64312dc) | [after](https://github.com/elastic/kibana/commit/9806b41fa9d260b42f7ec81e8927caeadf949066) | diff |
| --- | --- | --- | --- |
| `securitySolution` | 9.4MB | 9.4MB | +1.3KB |

#### Public APIs missing exports
> Total count of every type that is part of your API that should be exported but is not. This will cause broken links in the API documentation system. Target amount is 0. `Run node scripts/build_api_docs --plugin [yourplugin] --stats exports` for more detailed information.


| id | [before](https://github.com/elastic/kibana/commit/32e337401350aa1c34a30ceacc92b642d64312dc) | [after](https://github.com/elastic/kibana/commit/9806b41fa9d260b42f7ec81e8927caeadf949066) | diff |
| --- | --- | --- | --- |
| `alerting` | 54 | 57 | +3 |



### History
* :broken_heart: [Build #309981](https://buildkite.com/elastic/kibana-pull-request/builds/309981) failed 5ad4df4d9e42085fcfd6daaf1be28765bb5fd9d5
* :broken_heart: [Build #309954](https://buildkite.com/elastic/kibana-pull-request/builds/309954) failed d9177f09940603f0b41762f3ece68ed8d86b0f5f
* :broken_heart: [Build #309940](https://buildkite.com/elastic/kibana-pull-request/builds/309940) failed 48d16fff4f81298f0c8680d923faf600f0f84771
* :broken_heart: [Build #309881](https://buildkite.com/elastic/kibana-pull-request/builds/309881) failed 83e6af48c11b84960577445324bafeac245319c8
* :broken_heart: [Build #309731](https://buildkite.com/elastic/kibana-pull-request/builds/309731) failed f571801f381aa00fcfb651c6ab1f1f49acb09153

cc @denar50

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/310058","number":310058,"commit":"9806b41fa9d260b42f7ec81e8927caeadf949066"},{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/309981","number":309981,"commit":"5ad4df4d9e42085fcfd6daaf1be28765bb5fd9d5"},{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/309954","number":309954,"commit":"d9177f09940603f0b41762f3ece68ed8d86b0f5f"},{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/309940","number":309940,"commit":"48d16fff4f81298f0c8680d923faf600f0f84771"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/309881","number":309881,"commit":"83e6af48c11b84960577445324bafeac245319c8"},{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/309731","number":309731,"commit":"f571801f381aa00fcfb651c6ab1f1f49acb09153"}],"number":310058}
buildkite-pr-comment-->

marshallmain commented on x-pack/platform/plugins/shared/alerting/server/application/rule/methods/bulk_fill_gaps_by_rule_ids/process_gaps_batch.test.ts:
```diff
@@ -0,0 +1,168 @@
+      createGap([range('2025-05-09T09:15:09.457Z', '2025-05-11T09:15:09.457Z')]),
+      createGap([range('2025-05-11T09:15:09.457Z', '2025-05-17T09:15:09.457Z')]),
+    ];
+    const clappedGapsBatch = [
```
```suggestion
    const clampedGapsBatch = [
```

marshallmain reviewed (approved):
🚀

denar50 committed:
fix typo

denar50 committed:
Merge branch 'main' into security-team-10688-bulk-fill-gaps-for-rule-be

elasticmachine commented:
## :green_heart: Build Succeeded
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/310148)
* Commit: 88a1bd30313bcaa8524c667cd639cbc6bf4be161

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
#### Async chunks
> Total size of all lazy-loaded chunks that will be downloaded as the user navigates the app


| id | [before](https://github.com/elastic/kibana/commit/fa214dcf1c925bdda7b8a18f2acdef90ebd9ca0f) | [after](https://github.com/elastic/kibana/commit/88a1bd30313bcaa8524c667cd639cbc6bf4be161) | diff |
| --- | --- | --- | --- |
| `securitySolution` | 9.4MB | 9.4MB | +1.3KB |

#### Public APIs missing exports
> Total count of every type that is part of your API that should be exported but is not. This will cause broken links in the API documentation system. Target amount is 0. `Run node scripts/build_api_docs --plugin [yourplugin] --stats exports` for more detailed information.


| id | [before](https://github.com/elastic/kibana/commit/fa214dcf1c925bdda7b8a18f2acdef90ebd9ca0f) | [after](https://github.com/elastic/kibana/commit/88a1bd30313bcaa8524c667cd639cbc6bf4be161) | diff |
| --- | --- | --- | --- |
| `alerting` | 54 | 57 | +3 |



### History
* :green_heart: [Build #310058](https://buildkite.com/elastic/kibana-pull-request/builds/310058) succeeded 9806b41fa9d260b42f7ec81e8927caeadf949066
* :broken_heart: [Build #309981](https://buildkite.com/elastic/kibana-pull-request/builds/309981) failed 5ad4df4d9e42085fcfd6daaf1be28765bb5fd9d5
* :broken_heart: [Build #309954](https://buildkite.com/elastic/kibana-pull-request/builds/309954) failed d9177f09940603f0b41762f3ece68ed8d86b0f5f
* :broken_heart: [Build #309940](https://buildkite.com/elastic/kibana-pull-request/builds/309940) failed 48d16fff4f81298f0c8680d923faf600f0f84771
* :broken_heart: [Build #309881](https://buildkite.com/elastic/kibana-pull-request/builds/309881) failed 83e6af48c11b84960577445324bafeac245319c8
* :broken_heart: [Build #309731](https://buildkite.com/elastic/kibana-pull-request/builds/309731) failed f571801f381aa00fcfb651c6ab1f1f49acb09153

cc @denar50

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/310148","number":310148,"commit":"88a1bd30313bcaa8524c667cd639cbc6bf4be161"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/310058","number":310058,"commit":"9806b41fa9d260b42f7ec81e8927caeadf949066"},{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/309981","number":309981,"commit":"5ad4df4d9e42085fcfd6daaf1be28765bb5fd9d5"},{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/309954","number":309954,"commit":"d9177f09940603f0b41762f3ece68ed8d86b0f5f"},{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/309940","number":309940,"commit":"48d16fff4f81298f0c8680d923faf600f0f84771"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/309881","number":309881,"commit":"83e6af48c11b84960577445324bafeac245319c8"}],"number":310148}
buildkite-pr-comment-->

kibanamachine commented:
Starting backport for target branches: 8.19

https://github.com/elastic/kibana/actions/runs/15743888365

kibanamachine commented:
## 💔 All backports failed

| Status | Branch | Result |
|:------:|:------:|:------|
|❌|8.19|**Backport failed because of merge conflicts**<br><br>You might need to backport the following PRs to 8.19:<br> - [Optimize bulk actions endpoint & update gaps (#222158)](https://github.com/elastic/kibana/pull/222158)|

### Manual backport
To create the backport manually run:
```
node scripts/backport --pr 220866
```

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sorenlouv/backport)

<!--- Backport version: 9.6.6 -->

kibanamachine commented:
Friendly reminder: Looks like this PR hasn’t been backported yet.
To create automatically backports add a `backport:*` label or prevent reminders by adding the `backport:skip` label.
You can also create backports manually by running `node scripts/backport --pr 220866` locally
cc: @denar50
